### PR TITLE
Improve Base64 decoding performance

### PR DESCRIPTION
**Pull request by BlobCodes**
Supersedes #11032 

This PR improves the performance of Base64 decoding.
It replaces the previous PR to ease the reviewing of this PR.
It does not include the bang variants.

Benchmarks:
```
Benchmark: base64 decode (8 characters)
old  48.03M ( 20.82ns) (± 1.74%)  16.0B/op   1.06× slower
new  50.82M ( 19.68ns) (± 1.78%)  16.0B/op        fastest

Benchmark: base64 decode (8000 characters)
old 180.71k (  5.53µs) (± 1.34%)  5.88kB/op   1.17× slower
new 211.70k (  4.72µs) (± 1.24%)  5.88kB/op        fastest

Benchmark: base64 decode (80000000 characters)
old  20.35  ( 49.13ms) (± 0.30%)  57.2MB/op   1.19× slower
new  24.31  ( 41.13ms) (± 1.08%)  57.2MB/op        fastest
```

### CODE REVIEW

File: src/base64.cr [hunk 1]
```diff
@@ -244,65 +244,96 @@ module Base64
...
+      %buffer = (%buffer << 6) + %decoded
+      {% if raise_error == true %}
+        raise Base64::Error.new("Unexpected byte 0x#{{{byte}}.to_s(16)} at #{{{chunk_pos}} + {{i}}}") if %decoded == 255_u8
+      {% end %}
+    {% end %}
+
+    # Each byte in the buffer is shifted to rightmost position of the buffer, then casted to a UInt8
+    {% for i in 2..(bytes.size) %}
+      byte = (%buffer >> {{ (4 - bytes.size) * 2 + (8 * (bytes.size - i)) }}).to_u8!
+      {{ yield }}
```
[the remark this answers is not in the document]
Comment (kostya) [reply src/base64.cr:325]: I not sure, code became more magical then. now all explicit here.

[the remark this answers is not in the document]
Comment (BlobCodes) [reply src/base64.cr:325]: I also think that the macro yield is a bit more explicit

[the remark this answers is not in the document]
Comment (straight-shoota) [reply src/base64.cr:325]: Well, block arguments on macro blocks are already magically nonfunctionin and it only works because `byte` is assigned to a local variable in the macro.
I think it's perfectly fine to have the macro yield directly. It could be renamed to `yield_decoded_chunk_bytes` to make this explicit.

[the remark this answers is not in the document]
Comment (BlobCodes) [reply src/base64.cr]: The name `yield_decoded_chunk_bytes` is indeed quite explicit.
I have changed it now, thanks.

File: src/base64.cr [hunk 2]
```diff
@@ -244,65 +244,89 @@ module Base64
...
+    {% end %}
+
+    # Each byte in the buffer is shifted to rightmost position of the buffer, then casted to a UInt8
+    {% for i in 2..(bytes.size) %}
+      yield (%buffer >> {{ (4 - bytes.size) * 2 + (8 * (bytes.size - i)) }}).to_u8!
+    {% end %}
   end
 
-  private DECODE_TABLE = Array(Int8).new(256) do |i|
+  private DECODE_TABLE = StaticArray(UInt8, 256).new do |i|
```
Comment (straight-shoota) [inline src/base64.cr:322]: What's the reason for this change from `Array` to `StaticArray`?

File: src/base64.cr [hunk 3]
```diff
@@ -244,65 +244,89 @@ module Base64
...
-      a, b = next_decoded_value, next_decoded_value
-      yield (a << 2 | b >> 4).to_u8!
-    elsif mod == 3
-      a, b, c = next_decoded_value, next_decoded_value, next_decoded_value
-      yield (a << 2 | b >> 4).to_u8!
-      yield (b << 4 | c >> 2).to_u8!
-    elsif mod != 0
-      raise Error.new("Wrong size")
+    # If the amount of base64 characters is not divisable by 4, the remainder of the previous loop is handled here
+    unread_bytes = (fin - bytes) % 4
```
Comment (straight-shoota) [inline src/base64.cr:283]: Why is there a modulo operation? How can `unread_bytes` be `> 3`?

File: src/base64.cr [hunk 4]
```diff
@@ -244,65 +244,89 @@ module Base64
...
+  # The second argument (chunk_pos) is only used for the resulting error message.
+  # The resulting bytes are then each yielded to the given block, using the variable `byte`.
+  private macro yield_decoded_chunk_bytes(raise_error, chunk_pos, bytes, &block)
+    %buffer = 0_u32
+    {% for byte, i in bytes %}
+      %decoded = DECODE_TABLE.unsafe_fetch({{byte}})
+      %buffer = (%buffer << 6) + %decoded
+      {% if raise_error == true %}
+        raise Base64::Error.new("Unexpected byte 0x#{{{byte}}.to_s(16)} at #{{{chunk_pos}} + {{i}}}") if %decoded == 255_u8
+      {% end %}
```
Comment (straight-shoota) [inline src/base64.cr:313]: All calls to this method have `raise_error: true`, so it seems this macro conditional and the entire parameter are unnecessary.

File: src/base64.cr [hunk 5]
```diff
@@ -244,65 +244,89 @@ module Base64
...
-    cstr += 1
-    if res < 0
-      raise Error.new("Unexpected byte 0x#{sym.to_s(16)} at #{cstr - start_cstr - 1}")
-    end
-    res
+  # This macro decodes the given chunk of (2-4) base64 characters.
+  # The first argument (raise_error) determines if the macro raises an error when an invalid character is processed.
+  # The second argument (chunk_pos) is only used for the resulting error message.
+  # The resulting bytes are then each yielded to the given block, using the variable `byte`.
+  private macro yield_decoded_chunk_bytes(raise_error, chunk_pos, bytes, &block)
```
Comment (straight-shoota) [inline src/base64.cr:306]: This is really a minor detail, but I think it's easier to understand this macro if the `bytes` parameter comes first. Because those bytes are what's yielded. Makes sense to put that up front as a positional argument.

The `raise_error` parameter seems unnecessary because all calls pass `true`. The block param is also obsolete.

`bytes` could really be a splat arg. That fits very well:
```suggestion
  private macro yield_decoded_chunk_bytes(*bytes, chunk_pos)
```

Call site looks like this:

```cr
yield_decoded_chunk_bytes(bytes[0], bytes[1], bytes[2], chunk_pos: bytes - bytes_begin)
```

File: src/base64.cr (hunk 4 above)
Comment (BlobCodes) [reply src/base64.cr:313]: This was from the previous bang methods
I'll remove this later

File: src/base64.cr (hunk 2 above)
Comment (BlobCodes) [reply src/base64.cr:322]: Performance.
I need to benchmark this later, but it definitely gave a performance boost.

Comment (BlobCodes) [reply src/base64.cr:322]: Results:
```
staticarray 363.28k (  2.75µs) (± 4.41%)  2.42kB/op   1.06× slower
      array 384.81k (  2.60µs) (± 4.90%)  2.42kB/op        fastest
```

File: src/base64.cr (hunk 3 above)
Comment (BlobCodes) [reply src/base64.cr:283]: https://github.com/crystal-lang/crystal/blob/3b6e62c649f7097dec4b1e8785a6c4ea0f19726f/src/base64.cr#L275

Because it's been that way before.
I think it cannot be greater than 3, but I would have to double-check that.

Comment (BlobCodes) [reply src/base64.cr:279]: Just tested it, using `fin - bytes + 4` would also work.
the results of `fin - bytes` are in the range `-4..-1`.
I don't know if it's worth changing it.

File: src/base64.cr (hunk 2 above)
Comment (BlobCodes) [reply src/base64.cr:308]: Is the 6% improvement worth the 256 byte stack space or should it be changed back?

File: src/base64.cr (hunk 3 above)
Comment (straight-shoota) [reply src/base64.cr:279]: Ah, right. The difference is negative. I suppose the modulo operation is just used to make it positive, then.

File: src/base64.cr (hunk 2 above)
Comment (straight-shoota) [reply src/base64.cr:308]: I'm wondering where this difference is coming from... @asterite any thoughts what we should do here?

Comment (asterite) [reply src/base64.cr:308]: Ideally, have a syntax for const arrays and declare this as such, so that it goes in the ROM section. Given that we don't have that, it doesn't matter.

**Review by straight-shoota — APPROVED** [review]

**Review by oprypin — COMMENTED** [review]
I have only minor comments

File: src/base64.cr (hunk 2 above)
Comment (BlobCodes) [reply src/base64.cr:308]: I actually didn't notice that when writing the comment.
I'll turn it back to an array later.

File: src/base64.cr [hunk 6]
```diff
@@ -244,65 +244,75 @@ module Base64
...
-  private macro next_decoded_value
-    sym = cstr.value
-    res = dt[sym]
-    cstr += 1
-    if res < 0
-      raise Error.new("Unexpected byte 0x#{sym.to_s(16)} at #{cstr - start_cstr - 1}")
-    end
-    res
+  # This macro decodes the given chunk of (2-4) base64 characters.
+  # The first argument (raise_error) determines if the macro raises an error when an invalid character is processed.
```
[the remark this answers is not in the document]
Comment (BlobCodes) [reply src/base64.cr:291]: Ah right, that was removed after removing the bang variants.

File: src/base64.cr [hunk 7]
```diff
@@ -244,65 +244,75 @@ module Base64
...
-    when 'a'..'z' then (i - 0x47).to_i8
-    when '0'..'9' then (i + 0x04).to_i8
-    when '+', '-' then 0x3E_i8
-    when '/', '_' then 0x3F_i8
-    else               -1_i8
+    when 'A'..'Z' then (i - 0x41).to_u8
+    when 'a'..'z' then (i - 0x47).to_u8
+    when '0'..'9' then (i + 0x04).to_u8
+    when '+', '-' then 0x3E_u8
+    when '/', '_' then 0x3F_u8
```
[the remark this answers is not in the document]
Comment (BlobCodes) [reply src/base64.cr:314]: That will probably only improve the startup time by a few nanoseconds, but why not.

[the remark this answers is not in the document]
Comment (oprypin) [reply src/base64.cr]: Oh yeah, so now I fell for this myself :sweat_smile: 
Startup only, you're right!

**Comment by BlobCodes** [comment]
It seems like the last commit had two bangs which didn't belong there.. I'll fix that tomorrow

**Review by oprypin — APPROVED** [review]

**Review by straight-shoota — APPROVED** [review]
